### DISCUSSION TITLE
Using shared array (to represent matrix) in Hastlayer

### DISCUSSION THREAD

**Discussion by OlegKarasik** [Lombiq/Hastlayer-SDK#113] [Q&A]
Good morning,

I have a question about usage of shared array (to represent a matrix) with Hastlayer to write a parallel algorithm. 

Right now, I am trying to write a simple, parallel implementation of the Floyd-Warshall algorithm (you can find the source code [here](https://github.com/OlegKarasik/apsp-problems/blob/master/floyd-warshall/src/Hardware/APSP04.cs)) using Hastlayer. I have read the documentation and browsed through samples and as I understood the `SimpleMemory` class is a way to pass data to the FPGA unit. However, I still can't understand how to use it for this purpose, especially taking in account an explicit statement in documentation about `SimpleMemory` not being thread safe.

I am sure I am missing something conceptual here, so any help will be much appreciated.

**Comment by Piedone** [comment]
No problem, let me include an example: I modified the ParallelAlgorithm sample (https://github.com/Lombiq/Hastlayer-SDK/blob/dev/src/Samples/Hast.Samples.SampleAssembly/ParallelAlgorithm.cs) to give you an idea, see here: https://github.com/Lombiq/Hastlayer-SDK/commit/a4519e7a021b3d8f03eb3544e4d73dab48dc180f. I actually also discovered and fixed a bug in Hastlayer (https://github.com/Lombiq/Hastlayer-SDK/issues/110) for this; please update to the v2.1.1-alpha.0.hast-321 prerelease package to use the bug fix.

`ImageContrastModifier` (https://github.com/Lombiq/Hastlayer-SDK/blob/dev/src/Samples/Hast.Samples.SampleAssembly/ImageContrastModifier.cs) is also an example that's related, since it works with images, i.e. input data in a matrix form. It doesn't use arrays though, rather, it reads pixels out of the memory one by one. This may seem ridiculously inefficient at first, but Hastlayer actually accesses the memory with a cache with burst read/writes: it always utilizes the full bus width of the available RAM hardware, e.g. read/writing 512b at once, despite the consumer code seemingly reading/writing individual 32b values.

**Reply by Piedone** [reply]
In https://github.com/Lombiq/Hastlayer-SDK/pull/107 we fixed a couple of things too, so please use the latest v2.1.1-alpha.0.hast-321 package. We'll do a non-alpha release later too, once https://github.com/Lombiq/Hastlayer-SDK/issues/114 is fixed.
